### PR TITLE
Switch decorators to Rails builtin decorations - take 2

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

After we merged https://github.com/thepracticaldev/dev.to/pull/6040 we noticed we forgot to test serialization in decorated objects and thus unfortunately broke the website.

This PR adds back serialization to decorators, tests each decorator's serialization and the default home feed's serialization

## Related Tickets & Documents

https://github.com/thepracticaldev/dev.to/pull/6040
https://github.com/thepracticaldev/dev.to/pull/6122

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

### CODE REVIEW

File: spec/decorators/article_decorator_spec.rb [hunk 1]
```diff
@@ -6,7 +6,37 @@ def create_article(*args)
     article.decorate
   end
 
-  let(:article) { build_stubbed(:article) }
+  let(:article) { build(:article) }
+  let(:published_article) { create_article(published: true) }
+  let(:organization) { build(:organization) }
+
+  context "with serialization" do
```
Comment (rhymes) [inline spec/decorators/article_decorator_spec.rb]: new tests

File: spec/decorators/comment_decorator_spec.rb [hunk 2]
```diff
@@ -1,26 +1,42 @@
 require "rails_helper"
 
 RSpec.describe CommentDecorator, type: :decorator do
+  context "with serialization" do
```
Comment (rhymes) [inline spec/decorators/comment_decorator_spec.rb]: new tests

File: spec/decorators/notification_decorator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
+require "rails_helper"
+
+RSpec.describe NotificationDecorator, type: :decorator do
+  context "with serialization" do
```
Comment (rhymes) [inline spec/decorators/notification_decorator_spec.rb]: new tests

File: spec/decorators/organization_decorator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
+require "rails_helper"
+
+RSpec.describe OrganizationDecorator, type: :decorator do
+  context "with serialization" do
```
Comment (rhymes) [inline spec/decorators/organization_decorator_spec.rb]: new tests

File: spec/decorators/podcast_episode_decorator_spec.rb [hunk 5]
```diff
@@ -1,10 +1,80 @@
 require "rails_helper"
 
 RSpec.describe PodcastEpisodeDecorator, type: :decorator do
+  context "with serialization" do
```
Comment (rhymes) [inline spec/decorators/podcast_episode_decorator_spec.rb]: new tests

File: spec/decorators/sponsorship_decorator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
+require "rails_helper"
+
+RSpec.describe SponsorshipDecorator, type: :decorator do
+  context "with serialization" do
```
Comment (rhymes) [inline spec/decorators/sponsorship_decorator_spec.rb]: new tests

File: spec/decorators/user_decorator_spec.rb [hunk 7]
```diff
@@ -1,48 +1,107 @@
 require "rails_helper"
 
 RSpec.describe UserDecorator, type: :decorator do
-  let(:user) { build_stubbed(:user) }
+  let_it_be_changeable(:saved_user) { create(:user) }
+  let(:user) { build(:user) }
+
+  context "with serialization" do
```
Comment (rhymes) [inline spec/decorators/user_decorator_spec.rb]: new tests

File: spec/requests/stories_index_spec.rb [hunk 8]
```diff
@@ -68,6 +68,31 @@
...
     end
+
+    describe "when authenticated" do
+      let(:user) { create(:user) }
+
+      before do
+        sign_in user
+      end
+
+      it "contains the stories correctly serialized" do
```
Comment (rhymes) [inline spec/requests/stories_index_spec.rb]: testing the home feed!

File: spec/decorators/comment_decorator_spec.rb [hunk 9]
```diff
@@ -1,26 +1,42 @@
...
+    let_it_be_readonly(:comment) { create(:comment).decorate }
+
+    it "serializes both the decorated object IDs and decorated methods" do
+      expected_result = { "id" => comment.id, "published_timestamp" => comment.published_timestamp }
+      expect(comment.as_json(only: [:id], methods: [:published_timestamp])).to eq(expected_result)
```
Comment (mstruve) [inline spec/decorators/comment_decorator_spec.rb]: Do we test serializing all of the attribute methods `as_json` anywhere?

File: app/controllers/stories_controller.rb [hunk 10]
```diff
@@ -1,4 +1,17 @@
 class StoriesController < ApplicationController
+  DEFAULT_HOME_FEED_ATTRIBUTES_FOR_SERIALIZATION = {
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply app/controllers/stories_controller.rb]: ...how did you get this past CodeClimate? It had a fit whenever I tried to add to this file due to length!

File: app/views/stories/_main_stories_feed.html.erb [hunk 11]
```diff
@@ -2,15 +2,8 @@
...
-      <%= @stories.to_json(
-            only: %i[title path id user_id comments_count positive_reactions_count organization_id
-                     reading_time video_thumbnail_url video video_duration_in_minutes language
-                     experience_level_rating experience_level_rating_distribution cached_user cached_organization],
-            methods: %i[readable_publish_date cached_tag_list_array flare_tag class_name
-                        cloudinary_video_url video_duration_in_minutes published_at_int
-                        published_timestamp],
-          ) %>">
+    <div id="home-articles-object"
+         data-articles="<%= @stories.to_json(controller.class.const_get(:DEFAULT_HOME_FEED_ATTRIBUTES_FOR_SERIALIZATION)) %>">
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply app/views/stories/_main_stories_feed.html.erb]: Oh is that how you reference a constant in the Controller from the View?!

File: spec/decorators/comment_decorator_spec.rb (hunk 9 above)
Comment (rhymes) [reply spec/decorators/comment_decorator_spec.rb]: Not all the attributes, I replicated most examples using the various parameters. We do test the serialization of the homepage here: https://github.com/thepracticaldev/dev.to/pull/6125/files#diff-786f309cff05f0950d73a2b25f188fb1R79

Comment (mstruve) [reply spec/decorators/comment_decorator_spec.rb]: Oh nice bc if serializing a field was broken that would return a 500 👍

**Comment by rhymes** [comment]
@joshpuetz @mstruve @maestromac as I had to force push I'm re-requesting reviews.

BTW I think we should re-enable the automatic invalidation of approved PRs :D

File: app/decorators/user_decorator.rb [hunk 12]
```diff
@@ -18,14 +18,12 @@ class UserDecorator < ApplicationDecorator
...
-
   def cached_followed_tags
     Rails.cache.fetch("user-#{id}-#{updated_at}/followed_tags_11-30", expires_in: 20.hours) do
-      follows_query = Follow.where(follower_id: id, followable_type: "ActsAsTaggableOn::Tag").pluck(:followable_id, :points)
-      tags = Tag.where(id: follows_query.map { |f| f[0] }).order("hotness_score DESC")
+      follows = Follow.where(follower_id: id, followable_type: "ActsAsTaggableOn::Tag").pluck(:followable_id, :points)
+      tags = Tag.where(id: follows.map { |f| f[0] }).order("hotness_score DESC")
       tags.each do |t|
-        follow_query_item = follows_query.detect { |f| f[0] == t.id }
+        follow_query_item = follows.detect { |f| f[0] == t.id }
```
Comment (mstruve) [inline app/decorators/user_decorator.rb]: If we made follows a hash with pairs like this `{id: points}` then this lookup with be a lot faster bc we wouldn't have to iterate through the list every single time.

File: app/decorators/article_decorator.rb [hunk 13]
```diff
@@ -26,26 +24,29 @@ def url
...
       "medium"
     else
       "short"
     end
   end
 
   def internal_utm_params(place = "additional_box")
+    org_slug = organization&.slug
+
     campaign = if boosted_additional_articles
```
[the remark this answers is not in the document]
Comment (rhymes) [reply app/decorators/article_decorator.rb]: I had this exchange with @vaidehijoshi in the previous PR: https://github.com/thepracticaldev/dev.to/pull/6040#discussion_r378947643

I don't want to change the behavior as I don't know exactly what ripercussions it has, maybe someone can do it in a following PR

File: app/decorators/user_decorator.rb (hunk 12 above)
Comment (rhymes) [reply app/decorators/user_decorator.rb]: Done! https://github.com/thepracticaldev/dev.to/pull/6125/commits/790ecd29d0ac4b1ecf45e7a34b04f8bbb9b2df6d

File: app/decorators/user_decorator.rb [hunk 14]
```diff
@@ -18,15 +18,14 @@ class UserDecorator < ApplicationDecorator
...
-
   def cached_followed_tags
     Rails.cache.fetch("user-#{id}-#{updated_at}/followed_tags_11-30", expires_in: 20.hours) do
-      follows_query = Follow.where(follower_id: id, followable_type: "ActsAsTaggableOn::Tag").pluck(:followable_id, :points)
-      tags = Tag.where(id: follows_query.map { |f| f[0] }).order("hotness_score DESC")
-      tags.each do |t|
-        follow_query_item = follows_query.detect { |f| f[0] == t.id }
-        t.points = follow_query_item[1]
+      follows = Follow.where(follower_id: id, followable_type: "ActsAsTaggableOn::Tag").pluck(:followable_id, :points)
+      follows_map = follows.to_h
```
Comment (mstruve) [inline app/decorators/user_decorator.rb]: AH I always forget about the simple `.to_h` so slick! 🔥

**Comment by joshpuetz** [comment]
@rhymes We'll have to agree to disagree about clearing reviews on a force push: it personally drives me crazy as rebasing is how I like to catch a long running branch up with maste.r

**Comment by rhymes** [comment]
@joshpuetz maybe there's a half setting somewhere, as there's a possibility that external contributors add commits after we review them and don't notice, but we'll get there if and when we get there :)

I'm going to fix the conflict

**Pull request merged**